hail/src/main/resources/include/hail/NDArray.h

```diff
@@ -124,4 +124,40 @@ std::vector<long> unify_shapes(std::vector<long> &left, std::vector<long> &right
   return result;
 }
 
+// Assumption: left and right are of dimension 1 or more
+std::vector<long> matmul_shape(std::vector<long> &left, std::vector<long> &right) {
+  std::vector<long> result;
```

Comment:
I think we should prefer placing this as close to when it's needed as possible, so right before the if.